geoopt/manifolds/birkhoff_polytope.py

```diff
@@ -0,0 +1,143 @@
+import torch
+
+from geoopt import linalg
+from .base import Manifold
+
+__all__ = ["BirkhoffPolytope"]
+
+_birkhoff_doc = r"""
+    Manifold induced by the Doubly Stochastic matrices as described in 
+    A. Douik and B. Hassibi, "Manifold Optimization Over the Set 
+    of Doubly Stochastic Matrices: A Second-Order Geometry"
+    ArXiv:1802.02628, 2018.
+    Link to the paper: https://arxiv.org/abs/1802.02628.
+
+    @Techreport{Douik2018Manifold,
+       Title   = {Manifold Optimization Over the Set of Doubly Stochastic 
+                  Matrices: {A} Second-Order Geometry},
+       Author  = {Douik, A. and Hassibi, B.},
+       Journal = {Arxiv preprint ArXiv:1802.02628},
+       Year    = {2018}
+    }
+    
+    Please also cite:
+    Tolga Birdal, Umut Şimşekli,
+    "Probabilistic Permutation Synchronization using the Riemannian Structure of the BirkhoffPolytope Polytope"
+    IEEE Conference on Computer Vision and Pattern Recognition, CVPR, 2019
+    Link to the paper: https://arxiv.org/abs/1904.05814
+"""
+
+
+class BirkhoffPolytope(Manifold):
+    __doc__ = r"""
+    {}
+    """.format(_birkhoff_doc)
+    name = "BirkhoffPolytope"
+
+    ## batch_size \times n \times n
+    ndim = 3
+
+    def __init__(self, maxiter=100, tol=1e-3, epsilon=1e-12):
+        self.maxiter = maxiter
+        self.tol = tol
+        self.epsilon = epsilon
+
+    def _check_shape(self, x, name):
+        dim_is_ok = x.dim() == self.ndim
```

Comment:
I guess this explicitly prevents broadcasting behavior, right?

Reply:
In other manifolds there is a convention to have an arbitrary number of batch dims. In most cases it should not be that hard to make operations broadcastable